### PR TITLE
fix for the issue #12239 (Angle bug for (d, m, s) tuple input (deprecate dms_to_degrees))

### PR DESCRIPTION
**Pull request by webbdays**
fix for the issue #12239 

Two solutions are proposed.
code for solution 1 is presented in this pull request.

### CODE REVIEW

File: astropy/coordinates/angle_formats.py [hunk 1]
```diff
@@ -412,13 +412,29 @@ def degrees_to_dms(d):
...
+    sign = 1.0;
     # determine sign
-    sign = np.copysign(1.0, d)
+    if (d!=0):
+        sign = np.copysign(1.0, d)
+    # code  for solution 1 :
+    if (((d==0)) and (m!=0)):
+        sign = np.copysign(1.0, m);
+    elif (((d==0) and (m==0)) and (s!=0)):
+        sign = np.copysign(1.0, s)
```
Comment (nstarman) [inline astropy/coordinates/angle_formats.py:437]: The if statement can be simplified by using `elif` to guarantee that previous statements are true.

```suggestion
    if d != 0:
        sign = np.copysign(1.0, d)
    elif m != 0:
        sign = np.copysign(1.0, m);
    elif s != 0:
        sign = np.copysign(1.0, s)
    else:
        sign = 1.0
```

Or even just
```python

    sign = np.copysign(1.0, d if d != 0 else (m if m != 0 else (s if s != 0 else 1.0)))
```

**Comment by nstarman** [comment]
@webbdays, thanks for the PR and for addressing the style issues. The CI is failing because there are extant **array** issues. If ``d``, ``m``, or ``s`` are arrays then a scalar check will fail. The correct sign should propagate element-wise, so I fear this current approach will not work. A vectorized approach is needed, e.g.

```python
d = np.atleast_1d(d)
m = np.atleast_1d(m)
s = np.atleast_1d(s)
dis0 = d == 0
dmis0 = dis0 & (m == 0)
sign = np.copysign(1, d)
sign[dis0] = np.copysign(1, m[dis0])  # fix d=0, m!=0
sign[dmis0] = np.copysign(1, s[dmis0])  # fix 
```
However I think there's a much better solution by just examining the sign of ``d + m/60 + s/3600``.
I'm not sure why https://github.com/astropy/astropy/blob/7a5ed41e23e037a9f9cf0d1e4358c93dea854bb0/astropy/coordinates/angle_formats.py#L441-L447 takes the absolute values without restoring the signs.
e.g. ``d = np.copysign(1, d) * np.floor(np.abs(d))`` and similarly for `m` and `s`.

**Comment by webbdays** [comment]
Thanks for your if else logic improvement code.

Can you please give some numerical example for d,m,s to be arrays.
I didn't understand that.

**Comment by webbdays** [comment]
@nstarman 
Do you mean 
'dms_tuple = [ [d1,m1,s1], [d2,m2,s2] ]'

**Comment by nstarman** [comment]
I'm looking at the CI, e.g. https://github.com/astropy/astropy/runs/4195800626?check_suite_focus=true
A number of tests are failing because the checks for d, m, s = 0 assume scalar input, but the function can also take vector input.

**Comment by webbdays** [comment]
Ok

**Comment by webbdays** [comment]
@nstarman 

Lets look at this senarios.

Senario 1:
Here (d,m,s) is a single angle measured clockwise/anti-clockwise, whose unit is degree.

E.g 1 : 
24.5 deg = 24 deg + 0.5 deg

To represent smaller angles < 1 deg , 
arcmin unit is there.
1 deg = 60 arcmin
Same for arcsec , 1 arcmin = 60 arcsec .
1 deg = 3600 arcsec .

so there is no meaning for 1deg - 3 arcmin + 2 arcsec , when angle is measured in clockwise/anti-clockwise.

If angle is measured in anti-clockwise all  d,m,s are/should be negative.
If clockwise , all should be positive.
Should be informed to user in documentation.

senario 2:
any of the d,m,s can be negative. And can also have fractional part.


Now, take the example.
Eg 2: when adding angles.
Two ways to do this. That is by senario 1 and senario 2 .

If senario 1 :
(1,2,3) + (0,-4,-2) = (1,-2,1) .
There is problem here as the d,m,s are not  according to senario 1.


So , I think it is better to have senario 2.

Sign of angle should be taken as the sign of (d + m/60.0 + s/3600.0).

So, it is better to just return  (d+ m/60.0 + s/3600.0) from  dms_to_degrees function as suggested by @mhvk in the comment of issue #12239 .

I think , the initial code in angle_formats.py was written based on senario 1. And they may have thought of adding two (d,m,s) angles (if such a situation arises) by converting
these to degrees and adding those angles in degrees. But it is not efficient.
Adding them in d,m,s format  and then converting to degrees is efficient.

Finally, I think senario 2 is better.

with this senario 2, even if d and m and s are arrays of same length.

**Comment by webbdays** [comment]
If senario 1 :
In adding example if result d,m,s is not according to senario 1 then convert it to degrees and the convert back to dms according to senario 1.

File: astropy/coordinates/angle_formats.py [hunk 2]
```diff
@@ -412,13 +412,22 @@ def degrees_to_dms(d):
...
+    d = np.atleast_1d(d)
+    m = np.atleast_1d(m)
+    s = np.atleast_1d(s)
+    dmss = np.array([d,m,s]).T
+
+    for dms in dmss:
+        # looking for first non zero
+        for i in dms:
+            if i!=0:
+                sign = np.copysign(1,i); break;
```
Comment (nstarman) [inline astropy/coordinates/angle_formats.py:429]: This is implementation only works for scalars

Let's look at the following example

```python
>>> d = np.array([  0.,  45., -45.])
>>> m = np.array([  0.,  30., -30.])
>>> s = np.array([ 0.,  0., -0.])
>>> for dms in np.array([d,m,s]).T:
...     print(dms)
...     for i in dms:
...         print(i)
...         if i != 0:
...             sign = np.copysign(1, i)
...             break
>>> print("sign =", sign)
[0. 0. 0.]
0.0
0.0
0.0
[45. 30.  0.]
45.0
[-45. -30.  -0.]
-45.0
sign = -1.0
```
So the only sign that gets recorded is the last one.
Also, in the case of d=m=s=0, this actually never even assigns a value to 'sign'!

```python
>>> for dms in dmss:
...     for i in dms:
...         if i != 0:
...             sign = np.copysign(1, i)
...             break
...     print("current sign =", sign)
NameError: name 'sign' is not defined
```

**Comment by nstarman** [comment]
@webbdays In my mind the overall sign should match that of the equivalent value in degrees.

**Comment by nstarman** [comment]
@webbdays, can you roll back the PR to before you merge ``main`` into this branch. We try not to do that as it makes backporting and other maintenance quite challenging. Instead, PRs should be *rebased* (https://docs.astropy.org/en/latest/development/workflow/development_workflow.html?highlight=rebase#how-to-rebase).
Thanks!

**Comment by webbdays** [comment]
> @webbdays, In my mind the overall sign should match that of the equivalent value in degrees.

Ha yes.
It should be that way.

**Comment by webbdays** [comment]
> @webbdays, can you roll back the PR to before you merge ``main`` into this branch. We try not to do that as it makes backporting and other maintenance quite challenging. Instead, PRs should be *rebased* (https://docs.astropy.org/en/latest/development/workflow/development_workflow.html?highlight=rebase#how-to-rebase).
> Thanks!

Ok

**Comment by webbdays** [comment]
> > @webbdays, In my mind the overall sign should match that of the equivalent value in degrees.
> 
> Ha yes.
> It should be that way.

If that is the way we proceed,
this pull request is not related to it.
Can we proceed with new pull request for that concept?

**Comment by nstarman** [comment]
@webbdays, either is fine: you can either edit this PR in-place (probably want to squash some commits) or close this PR and open a new one. Thanks for the sustained effort and engagement in helping us to close this bug!

**Comment by webbdays** [comment]
Can anyone please list the different places in which dms_to_degrees function is used in astropy.
Places I mean another functions or some class that use this function.

Is there any tool to get that list when required function is given as input ?

@nstarman

**Comment by eerovaher** [comment]
@webbdays, use [git-grep](https://git-scm.com/docs/git-grep).

**Comment by webbdays** [comment]
I used git grep to search.
Thanks @eerovaher .

I found that dms_to_degrees function is not much used in astropy. Except in Angle class.

Also, it is set to deprecate in upcoming v5.0 release.

**Comment by webbdays** [comment]
I have created a new branch named total_degrees.

I made the dms_degrees function to just return the required expression.
Removed minute and second range check. I think it is good to leave for user to decide.
May be user want to find the angle(degrees) for 90 arcmin.
(0, 90, 0)
It is same as (0,59, 31*60) .
The s= None handling code and error handling code are unchanged.

I going to make new pull request on total_degrees branch.

If so, I will close this pull request.

**Comment by nstarman** [comment]
@webbdays, if you have a new approach, it's fine to close this PR and submit a new one!

Just as a note, I think this PR can be rebased and squashed. The current solution feels a little convoluted, introducing a nested `for`-loop. Intrinsically vectorized functions can achieve faster results. I played around with the current code and I think the following is sufficient to solve the bug:  add some ``copysign`` to undo the ``np.abs`` operations and delete unnecessary ``np.abs``. Addition in the ``return`` statement takes care of the overall sign.

```python
    try:
        d = np.copysign(1.0, d) * np.floor(np.abs(d))
        if s is None:
            s = 0
        else:
            m = np.copysign(1.0, m) * np.floor(np.abs(m))
    except ValueError as err:
        raise ValueError(format_exception(
            "{func}: dms values ({1[0]},{2[1]},{3[2]}) could not be "
            "converted to numbers.", d, m, s)) from err
    return d + m / 60. + s / 3600.
```

`dms_to_degrees(0, -22, s=None)` -> -0.36666666666666664

I think removing the range protections should be approached in a separate issue / PR. This one should focus on closing #12239.

**Comment by webbdays** [comment]
Ok.
But but what is the need of floor.
Let the user to give fractional part as well.

Suppose the user want to find degrees for 2.5 arcmin.

**Comment by nstarman** [comment]
@mhvk, do you know what's happening in this function? The `floor`, and sign oddness is quite strange...

**Comment by mhvk** [comment]
@nstarman - see https://github.com/astropy/astropy/issues/12239#issuecomment-964345806 - my feeling is that this code is pretty broken to start with... Possibly the best solution is simply to deprecate and eventually remove it.

**Comment by pllim** [comment]
Given https://github.com/astropy/astropy/pull/12443#issuecomment-982050880 , I labeled this as `Close?` because it looks like we want to get rid of the function, not to fix it.

**Comment by webbdays** [comment]
I request to keep this function atleast for clock wise and anti-clockwise angles.
I think in most usage of angles these two are used frequently.
Arguments may look like this :
(dms_tuple, direction)
dms_tuple is (d,m,s).
direction is clockwise or anti-clockwise.

I think the user who raised the issue #12443 , seems to use this way either clockwise or anti-clockwise.

**Pull request closed without merging**